Update wc-sale-discord-notifications.php

Dextiz opened the pull request:
Version 2.1 – Major Feature & Stability Update
This release brings substantial upgrades to the WC Sale Discord Notifications plugin, including requested admin controls and robust safeguards against duplicate notifications.

~~ New Features ~~
* Admin-Controlled Notification Fields You now have full control over what data gets sent to Discord. Choose to include:

✅ Order Status
✅ Payment Info
✅ Product Details
✅ Creation Date
✅ Billing Information
✅ Transaction ID

This gives you granular privacy and clarity based on your business needs.

* Per-Status Customization Assign custom webhook URLs per WooCommerce order status


** New: Duplicate Notification Protection
We’ve solved the infamous “double ping” problem when customers refresh their thank-you page:

✅ A local discord_notification_log.txt is now used to track sent messages by order_id|type

✅ Ensures each order is only notified once per event type (new, update)

✅ Lightweight and file-based — no DB bloat, works with caching setups

🧰 Technical Enhancements
Declares WooCommerce 8+ compatibility with Custom Order Tables

Admin settings organized under WooCommerce > Discord Notifications

Adds settings shortcut link in plugin list

Fully localized and uses WordPress core features (register_setting, get_option, etc.)

🧪 Tested With
WordPress 6.6.2

WooCommerce 9.3.3

PHP 8.x

* File Added discord_notification_log.txt (auto-generated)
(Stores order tracking data to prevent duplicates)

Dextiz committed:
Update wc-sale-discord-notifications.php

Version 2.1 – Major Feature & Stability Update
This release brings substantial upgrades to the WC Sale Discord Notifications plugin, including requested admin controls and robust safeguards against duplicate notifications.

~~ New Features ~~
* Admin-Controlled Notification Fields
You now have full control over what data gets sent to Discord. Choose to include:

✅ Order Status
✅ Payment Info
✅ Product Details
✅ Creation Date
✅ Billing Information
✅ Transaction ID

This gives you granular privacy and clarity based on your business needs.

* Per-Status Customization
Assign custom webhook URLs per WooCommerce order status


** New: Duplicate Notification Protection
We’ve solved the infamous “double ping” problem when customers refresh their thank-you page:

✅ A local discord_notification_log.txt is now used to track sent messages by order_id|type

✅ Ensures each order is only notified once per event type (new, update)

✅ Lightweight and file-based — no DB bloat, works with caching setups

🧰 Technical Enhancements
Declares WooCommerce 8+ compatibility with Custom Order Tables

Admin settings organized under WooCommerce > Discord Notifications

Adds settings shortcut link in plugin list

Fully localized and uses WordPress core features (register_setting, get_option, etc.)

🧪 Tested With
WordPress 6.6.2

WooCommerce 9.3.3

PHP 8.x

* File Added
discord_notification_log.txt (auto-generated)
(Stores order tracking data to prevent duplicates)

Dextiz commented:
Now also tested on Woo 9,7,1 with about 100~ orders with no dupes.
WP 6,7,2, PHP 8,3,17

Cral-Cactus commented:
@Dextiz thank you for the support, I have been really busy last months and I couldn't do any updates, I will test this out tomorrow and push it!

Dextiz commented:
@Cral-Cactus have you been able to check it? :)

Dextiz commented:
Noticed it's not updated yet, I've run my version since 29th of March with 0 issues.

Cral-Cactus commented:
I've been really busy last months even now, I'll take a look once i get free time, now I requrested copilot lets see what it says

Cral-Cactus commented:
You can improve those things if you got time for that

Cral-Cactus commented:
@Dextiz let me know.

Dextiz committed:
Clean up duplicate log check, consolidate file writes, finalize v2.1

Cral-Cactus commented:
> Can I review this?

go for it

MogoEmo commented:
> @Dextiz @MogoEmo let me know

Please could I have the .zip to load it :)

Cral-Cactus commented:
> > @Dextiz @MogoEmo let me know
> 
> Please could I have the .zip to load it :)

@MogoEmo this is the last released zip which I have, https://github.com/Cral-Cactus/wc-sale-discord-notifications/releases/download/2.0/wc-sale-discord-notifications.zip , if you want to try the thing above, you have to edit the code and paste it in

Dextiz committed:
Refactor: Improve WC Sale Discord Notifications plugin (v2.2.0)

- Implemented per-status duplicate protection using order meta instead of global flag
- Removed redundant duplicate-check logic and double log writes
- Added sanitization callbacks for all plugin options to improve data safety
- Made Discord webhook POST asynchronous (blocking => false) with basic error handling
- Improved status change hook to only trigger on selected statuses
- Enhanced embed field building with formatted totals, safe hex color handling, and image fallback
- Updated “Tested up to” and “WC tested up to” versions
- Verified and reviewed changes with ChatGPT 5 for code quality and logic consistency

Dextiz committed:
Update README.md

Dextiz commented:
> @Dextiz @MogoEmo let me know

Took a while for me, been busy moving and summer. I've tried it on our store and so far 0 issues.

MogoEmo commented:
> @Dextiz @MogoEmo let me know

Love the update. No issues. Was able to add a bit of code so I could get notifications catered to my needs. In love with it. Thank you

Dextiz committed:
Bugg fix, always sends "new order" instead of "order updated" when someone cancels order at checkout and then pays.

Dextiz committed:
Did code review with Co-pilot and Gpt-5

-Debug option blocking http requests.
- Robust boolean sanitization for disable-image option; handle 'on'/1/0 consistently
- Ensure unchecked checkbox submits via hidden input; prevents sticky values in admin
- Add http args filter (wc_sale_discord_http_args) to toggle blocking mode for debugging
- Log HTTP errors when blocking is enabled; lightweight async attempt log otherwise
- Remove redundant ucwords() on wc_get_order_status_name()
- Optional: remove unused wc_sale_is_wc_order() helper
- Add PHPDoc to public methods and recommend CS rules for consistent formatting
- Double‑checked with ChatGPT 5 AND Co-pilot

SENPAY98K commented:
Does this support custom product fields ? 
When using plugins like `Advanced Product Fields (Product Addons) for WooCommerce`

Dextiz commented:
> Does this support custom product fields ? When using plugins like `Advanced Product Fields (Product Addons) for WooCommerce`

I don't see what would cause an issue. What the plugin does is send a notification to Discord once a order gets a new status. 
So should work without any issue.

<img width="1007" height="925" alt="chrome_VxJNUoLhxX" src="https://github.com/user-attachments/assets/308a2b59-6c7a-4a65-a7c8-f990069e4086" />

SENPAY98K commented:
I tried the this changes, and it does not show custom fields

Dextiz commented:
> I tried the this changes, and it does not show custom fields

Try the new one I publish in a few minutes and get back to me if it works.

Dextiz committed:
Feat: Add support for custom product fields in Discord notifications

- New toggle "Custom Fields" in settings
- When enabled, product-level custom fields (from addons/APF) are included in order item details

Dextiz commented:
<img width="601" height="164" alt="image" src="https://github.com/user-attachments/assets/344f9097-56c2-4c16-9955-15a3aaac0dcb" />
This will automatically include all APF/PAO/WC item meta under the product line, indented with a bullet.

It uses WooCommerce’s own "get_formatted_meta_data()" so it will(Should) work with Advanced Product Fields, Product Add-ons, and similar extensions.

Dextiz committed:
Update README.md

Dextiz committed:
Small styling requested by @Senpay98K

Dextiz committed:
Merge branch 'Dextiz-patch-1' of https://github.com/Dextiz/wc-sale-discord-notifications into pr/Dextiz/7

Cral-Cactus commented:
<img width="500" height="500" alt="logo" src="https://github.com/user-attachments/assets/93cbb015-266c-40a4-944a-cc58023e9c89" />
what do you guys think about this for new logo

Cral-Cactus commented:
gonna change the logo then tonight push it to wp

Cral-Cactus commented:
@Dextiz or @SENPAY98K do you guys think you can get me a screenshot like this one which i will attach below so i can update it on plugin page?
<img width="1119" height="815" alt="image" src="https://github.com/user-attachments/assets/fe49ad64-aa32-4892-8e72-b5e503a58f26" />

Cral-Cactus commented:
and maybe few screenshots of the discord webhooks?

SENPAY98K commented:
# WP / WC Menu

<img width="1214" height="963" alt="image1" src="https://github.com/user-attachments/assets/7d22ebd3-cc1b-48a9-90d2-b89d21cc3304" />

# Discord Embed

<img width="481" height="729" alt="image2" src="https://github.com/user-attachments/assets/b836ec9f-3a41-443d-96b5-bd8aef1e04ed" />

Cral-Cactus commented:
@SENPAY98K thank you, I will attach them on the plugin page!
